edgehog-device-runtime-forwarder/src/connections_manager.rs

```diff
@@ -0,0 +1,304 @@
+use displaydoc::Display;
+use futures::{future, SinkExt, StreamExt, TryFutureExt};
+use thiserror::Error as ThisError;
+use tokio::net::TcpStream;
+use tokio::select;
+use tokio::sync::mpsc::{channel, Receiver};
+use tokio_tungstenite::{
+    connect_async, tungstenite::Error as TungError, tungstenite::Message as TungMessage,
+    MaybeTlsStream, WebSocketStream,
+};
+use tracing::{debug, error, info, instrument, trace, warn};
+use url::Url;
+
+use crate::collection::Connections;
+use crate::connection::ConnectionError;
+use crate::messages::{
+    Http, HttpMessage, Id, ProtoMessage, Protocol as ProtoProtocol, ProtocolError,
+};
+
+/// Size of the channels where to send proto messages.
+pub(crate) const CHANNEL_SIZE: usize = 50;
+
+/// Errors occurring during the connections management.
+#[derive(Display, ThisError, Debug)]
+#[non_exhaustive]
+pub enum Error {
+    /// Error performing exponential backoff when trying to (re)connect with the bridge.
+    WebSocket(#[from] TungError),
+    /// Protobuf error.
+    Protobuf(#[from] ProtocolError),
+    /// Connection error.
+    Connection(#[from] ConnectionError),
+    /// Wrong message with id `{0}`
+    WrongMessage(Id),
+    /// The connection does not exists, id: `{0}`.
+    ConnectionNotFound(Id),
+    /// Connection ID already in use, id: `{0}`.
+    IdAlreadyUsed(Id),
+    /// Unsupported message type
+    Unsupported,
+    /// Session token not present on URL
+    TokenNotFound,
+    /// Session token already in use
+    TokenAlreadyUsed(String),
+    /// Error while performing exponential backoff to create a WebSocket connection
+    BackOff(#[from] BackoffError<Box<Error>>),
+}
+
+/// WebSocket stream alias.
+pub type WsStream = WebSocketStream<MaybeTlsStream<TcpStream>>;
+
+/// Handler responsible for establishing a websocket connection between a device and the bridge
+/// and for receiving and sending data from/to it.
+#[derive(Debug)]
+pub struct ConnectionsManager {
+    /// Collection of connections, each identified by an ID.
+    connections: Connections,
+    /// Websocket stream between the device and the bridge.
+    ws_stream: WsStream,
+    /// Channel used to send through the websocket messages coming from each connection.
+    rx_ws: Receiver<ProtoMessage>,
+    /// bridge URL.
+    url: Url,
+}
+
+impl ConnectionsManager {
+    /// Establish a new WebSocket connection between the device and the bridge.
+    #[instrument]
+    pub async fn connect(url: Url) -> Result<Self, Error> {
+        // TODO: check if, when a wrong URL is passed, it will endlessly try to connect
+
+        let ws_stream = Self::ws_connect(&url).await?;
+
+        // this channel is used by tasks associated to each connection to communicate new
+        // information available on a given websocket between the device and TTYD.
+        // it is also used to forward the incoming data from TTYD to the device.
+        let (tx_ws, rx_ws) = channel(CHANNEL_SIZE);
```

Comment:
So in this case the `CHANNEL_SIZE` is the number of open forwarding connections?

Reply:
Within a session established between the device and the bridge the `ConnectionsManager` uses this channel to receive data from the internal device connections (such as those with TTYD). A single forwarding connection may send more data on the channel, thus it is unlikely that CHANNEL_SIZE is effectively the same as the number of connections.